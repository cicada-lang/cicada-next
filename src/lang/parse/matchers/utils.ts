import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
