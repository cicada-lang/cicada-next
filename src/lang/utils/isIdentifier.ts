const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
