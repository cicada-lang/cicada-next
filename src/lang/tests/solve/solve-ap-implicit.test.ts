import { expect, test } from "vitest"
import { runCode } from "../utils"

test("solve ApImplicit -- same PatternVar", async () => {
  const output = await runCode(`

solve (id: (implicit T: Type, x: T) -> T, x: String) {
  unify id(x) = id(x)
}

`)

  expect(output).toMatchInlineSnapshot(
    '"{ id: TODO((implicit T2: Type, x1: T2) -> T2), x: TODO(String) }"',
  )
})

test("solve ApImplicit -- PatternVar v.s. String", async () => {
  const output = await runCode(`

solve (id: (implicit T: Type, x: T) -> T, x: String, A: Type) {
  unify id(implicit String, x) = id(implicit A, "abc")
}

`)

  expect(output).toMatchInlineSnapshot(
    '"{ id: TODO((implicit T: Type, x1: T) -> T), x: \\"abc\\", A: String }"',
  )
})

test("solve ApImplicit -- insertion", async () => {
  const output = await runCode(`

solve (id: (implicit T: Type, x: T) -> T, x: String) {
  unify id(x) = id("abc")
}

`)

  expect(output).toMatchInlineSnapshot(
    '"{ id: TODO((implicit T2: Type, x1: T2) -> T2), x: \\"abc\\" }"',
  )
})
