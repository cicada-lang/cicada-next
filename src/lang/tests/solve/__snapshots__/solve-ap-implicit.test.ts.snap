// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`solve ApImplicit -- PatternVar v.s. String > 
    "{
      id: ?id,
      x: \\"abc\\",
      A: String
    }"
   1`] = `
"{
  id: ?id,
  x: \\"abc\\",
  A: String
}"
`;

exports[`solve ApImplicit -- PatternVar v.s. String: 
    "{
      id: ?id,
      x: "abc",
      A: String
    }"
   1`] = `
"{
  id: ?id,
  x: "abc",
  A: String
}"
`;

exports[`solve ApImplicit -- deepWalk > 
    "{
      id: ?id,
      A: String,
      idA: ?id(implicit String)
    }"
   1`] = `
"{
  id: ?id,
  A: String,
  idA: ?id(implicit String)
}"
`;

exports[`solve ApImplicit -- deepWalk -- inserted > 
    "{
      id: ?id,
      x: \\"abc\\",
      c: ?id(implicit String, \\"abc\\")
    }"
   1`] = `
"{
  id: ?id,
  x: \\"abc\\",
  c: ?id(implicit String, \\"abc\\")
}"
`;

exports[`solve ApImplicit -- deepWalk -- inserted: 
    "{
      id: ?id,
      x: "abc",
      c: ?id(implicit String, "abc")
    }"
   1`] = `
"{
  id: ?id,
  x: "abc",
  c: ?id(implicit String, "abc")
}"
`;

exports[`solve ApImplicit -- deepWalk: 
    "{
      id: ?id,
      A: String,
      idA: ?id(implicit String)
    }"
   1`] = `
"{
  id: ?id,
  A: String,
  idA: ?id(implicit String)
}"
`;

exports[`solve ApImplicit -- insertion > 
    "{
      id: ?id,
      x: \\"abc\\"
    }"
   1`] = `
"{
  id: ?id,
  x: \\"abc\\"
}"
`;

exports[`solve ApImplicit -- insertion: 
    "{
      id: ?id,
      x: "abc"
    }"
   1`] = `
"{
  id: ?id,
  x: "abc"
}"
`;

exports[`solve ApImplicit -- same PatternVar > 
    "{
      id: ?id,
      x: ?x
    }"
   1`] = `
"{
  id: ?id,
  x: ?x
}"
`;

exports[`solve ApImplicit -- same PatternVar: 
    "{
      id: ?id,
      x: ?x
    }"
   1`] = `
"{
  id: ?id,
  x: ?x
}"
`;

exports[`solve ApImplicit -- same PatternVar 1`] = `
"{
  id: ?id,
  x: ?x
}"
`;

exports[`solve ApImplicit -- PatternVar v.s. String 1`] = `
"{
  id: ?id,
  x: "abc",
  A: String
}"
`;

exports[`solve ApImplicit -- insertion 1`] = `
"{
  id: ?id,
  x: "abc"
}"
`;

exports[`solve ApImplicit -- deepWalk 1`] = `
"{
  id: ?id,
  A: String,
  idA: ?id(implicit String)
}"
`;

exports[`solve ApImplicit -- deepWalk -- inserted 1`] = `
"{
  id: ?id,
  x: "abc",
  c: ?id(implicit String, "abc")
}"
`;
